src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -84,17 +80,41 @@ public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAl
 
         public virtual async Task WriteAsync(HubMessage message)
         {
+            await _writeLock.WaitAsync();
+
             try
             {
-                await _writeLock.WaitAsync();
+                // This will internally cache the buffer for each unique HubProtocol/DataEncoder combination
+                // So that we don't serialize the HubMessage for every single connection
+                var buffer = message.WriteMessage(ProtocolReaderWriter);
+                _connectionContext.Transport.Output.Write(buffer);
 
-                var buffer = ProtocolReaderWriter.WriteMessage(message);
+                Interlocked.Exchange(ref _lastSendTimestamp, Stopwatch.GetTimestamp());
 
-                _connectionContext.Transport.Output.Write(buffer);
+                await _connectionContext.Transport.Output.FlushAsync();
+            }
+            finally
+            {
+                _writeLock.Release();
+            }
+        }
+
+        private async Task TryWritePingAsync()
+        {
+            // Don't wait for the lock, if it returns false that means someone wrote to the connection
+            // and we don't need to send a ping anymore
+            if (!await _writeLock.WaitAsync(0))
+            {
+                return;
+            }
+
+            try
+            {
+                _connectionContext.Transport.Output.Write(_cachedPingMessage);
```

Comment:
Assert `_cachedPingMessage != null`. It depends upon the caller initializing properly so we should assert it.